File: Exercises/exercise02/main.py
from threading import Thread, Lock



i = 0
lock = Lock()

def Thread_func1():
	global i
	for j in range(1000000):
		lock.acquire()
		try:
		    i = i + 1
		finally:
		    lock.release()
		

		

		
def Thread_func2():
	global i
	for k in range(1000000):
		lock.acquire()
		try:
		    i = i - 1
		finally:
		    lock.release()


def main():
	
	Thread1 = Thread(target = Thread_func1(), args = (),)
	Thread2 = Thread(target = Thread_func2(), args = (),)
	Thread1.start()
	Thread2.start()
    
	Thread1.join()
	Thread2.join()
	
	print(i)
	

	
main()

File: Exercises/exercise02/main.c
//gcc -std=gnu99 -Wall -g -o main main.c -lpthread



#include <pthread.h>
#include <stdio.h>



int i = 0;
pthread_mutex_t lock;

void* Thread_func1(){
	int j;
	for(j = 0; j<1000000; j++){
		pthread_mutex_lock(&lock);
		i+=1;
		pthread_mutex_unlock(&lock);
	}
	
	
	return NULL;
}

void* Thread_func2(){
	int j;
	for(j = 0; j<1000000; j++){
		pthread_mutex_lock(&lock);
		i-=1;
		pthread_mutex_unlock(&lock);
	}
	 

	return NULL;
}



int main(){
    pthread_mutex_init(&lock, NULL);
    pthread_t Thread1, Thread2;
    pthread_create(&Thread1, NULL, Thread_func1, NULL);
    pthread_create(&Thread2, NULL, Thread_func2, NULL);
    pthread_join(Thread1, NULL);
    pthread_join(Thread2, NULL);
    printf("Value of i: %d \n", i);
	
	return 0;
}
